### ISSUE TITLE
Feature count regression on QGIS 3.16.X on views on windows

### ISSUE THREAD

**Issue by ponceta** [qgis/QGIS#41188]
**Describe the bug**
Feature count goes wrong on some views on QGIS 3.16.X

Works fine on same project on QGIS 3.10.X

QGIS 3.16.X comes again with annoying windows message for every feature count executed.

**How to Reproduce**

QGIS : 3.10.X

![image](https://user-images.githubusercontent.com/8614223/105856726-d6299480-5fe9-11eb-995c-64fe58785b72.png)

QGIS : 3.16.X

![image](https://user-images.githubusercontent.com/8614223/105856770-de81cf80-5fe9-11eb-915c-c482bae4ee73.png)

**QGIS and OS versions**

QGIS 3.16.3 Hannover Windows 10 x64

**Additional context**

Project issue : https://github.com/QGEP/QGEP/issues/627

**Comment by roya0045** [comment]
These are non-geometric data sources, its completely normal that no features are counted since they don't display any features.

**Comment by ponceta** [comment]
> These are non-geometric data sources, its completely normal that no features are counted since they don't display any features.

These are empty views, they should display [0] 

A select count (*) vw_qgep_maintenance in pgadmin shows 0 entities.

**Comment by roya0045** [comment]
If you import those layers in a new project, is the count still the same? Maybe the cached value was corrupted/altered.

**Comment by ponceta** [comment]
No, count is back to normal when reloading these views.

![image](https://user-images.githubusercontent.com/8614223/105861209-c2ccf800-5fee-11eb-9ccb-d6c385fb5ad2.png)

But reload the project and you get altered/corrupted values again 

![image](https://user-images.githubusercontent.com/8614223/105861225-c82a4280-5fee-11eb-9774-9761bf308aa4.png)

**Comment by roya0045** [comment]
Were the tables always empty or was the 90 and 80 count valid at some point?

**Comment by gioman** [comment]
> Project issue : [QGEP/QGEP#627](https://github.com/QGEP/QGEP/issues/627)

@ponceta that project cannot be opened, the layers are from a postgresql server that is not accessible remotely.

**Comment by ponceta** [comment]
No, they never had any entities. 

Ok on a blank project loading these layers don't end with the same result. :( That's a qgs project based issue! Erf...

Any idea where to look?

![image](https://user-images.githubusercontent.com/8614223/105872996-4ee51c80-5ffb-11eb-87da-941f31c30702.png)

**Comment by gioman** [comment]
@ponceta can you provide a dump (and a project) we can use to test on our machines?

**Comment by ponceta** [comment]
https://wetransfer.com/downloads/a5169232b2ecfa9d1cd1abb5508048dc20210126164605/eafa2223b2368971121c47617a0d7e8520210126164635/add638 

Data dump from pgadmin 4 and qgs project

**Comment by gioman** [comment]
> Data dump from pgadmin 4 and qgs project

@ponceta 

![Screenshot_20210126_175528](https://user-images.githubusercontent.com/1951107/105884343-cddc5400-5fff-11eb-9731-329f2a71370c.png)

![Screenshot_20210126_175605](https://user-images.githubusercontent.com/1951107/105884348-cf0d8100-5fff-11eb-9389-22e0930b9e64.png)

**Comment by ponceta** [comment]
If you look at the Dégâts layer which is also a view, it shows 80 entities but it is empty and should display [0].

But it seems that you have a different behaviour.

On my side I have 90 displayed as entity count for Maintenance view and 1 entity filtered.

![image](https://user-images.githubusercontent.com/8614223/105953754-01f15c80-6074-11eb-8b23-9d9eb06b7858.png)

**Comment by Gustry** [comment]
Linked to https://github.com/qgis/QGIS/issues/40162 ?

**Comment by gioman** [comment]
> Linked to #40162 ?

@Gustry I'm using master so it should contain that fix, and I still see the problem.

**Comment by troopa81** [comment]
@ponceta 

The estimated count for view was not working in 3.10 (see my comment [here](https://github.com/qgis/QGIS/pull/32114#issuecomment-538855734) ) and was returning always 0. I fixed it in #32114, and then fixed the following issue related to tables in #40162

The thing is, it's an estimation and it could be wrong. I use the *Plan rows* from EXPLAIN command to get the estimated count. For instance, on QGEP data, I have 

```sql
EXPLAIN (FORMAT JSON) SELECT 1 FROM qgep_od.infiltration_installation;
                     QUERY PLAN                      
-----------------------------------------------------
 [                                                  +
   {                                                +
     "Plan": {                                      +
       "Node Type": "Seq Scan",                     +
       "Parallel Aware": false,                     +
       "Relation Name": "infiltration_installation",+
       "Alias": "infiltration_installation",        +
       "Startup Cost": 0.00,                        +
       "Total Cost": 13.70,                         +
       "Plan Rows": 370,                            +
       "Plan Width": 4                              +
     }                                              +
   }                                                +
 ]
(1 ligne)

qgep_prod=# select count(*) from qgep_od.infiltration_installation; 
 count 
-------
     0
(1 ligne)
```

Even VACUUM ANALYSE won't change anything. I don't see any way to have a better estimation. 

I propose to close the ticket, except if somebody has a better idea on how to deal with this

**Comment by gioman** [comment]
> I don't see any way to have a better estimation.

@troopa81 is this a common problem to all clients (i.e. pgadmin, dbeaver, etc.)?

if yes, than I guess we should raise a **very clear** warning when enabling feature count on views.

**Comment by roya0045** [comment]
@gioman The other way would be to count and not estimate.

**Comment by troopa81** [comment]
> if yes, than I guess we should raise a very clear warning when enabling feature count on views.

The option is called *Use estimated table metadata*, so it's quite clear that the metadata will be estimated and not exact. But still maybe the user don't understand what metadata means (feature count, geometry type for some provider) and what mean estimated.

We can maybe improve the tooltip

**Comment by gioman** [comment]
> We can maybe improve the tooltip

@Gustry I'm afraid most of the users will overlook the tooltip, in this case I would see better a warning in the message bar.

**Comment by gioman** [comment]
> @gioman The other way would be to count and not estimate.

@roya0045 agree.

**Comment by troopa81** [comment]
> @Gustry I'm afraid most of the users will overlook the tooltip, in this case I would see better a warning in the message bar.

When you check the option ? or when you load the layers ? The second one seems a little bit too much but why not for the first one

**Comment by gioman** [comment]
@troopa81 yeah I guess that if would be enough when checking the option, after that is up to the user of course. But at the moment there isn't anything that clearly warns the users of the downsides of that option, while it is often suggested to activate it to "speed up" things.
